// vscode-typescript-languageservice/out/services/codeAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const vscode = require("vscode-languageserver");
const shared = require("@volar/shared");
const rename_1 = require("./rename");
const fixNames = require("../utils/fixNames");
function register(languageService, getTextDocument, host) {
    return async (uri, range, context) => {
        var _a, _b, _c, _d;
        const document = getTextDocument(uri);
        if (!document)
            return;
        const [formatOptions, preferences] = await Promise.all([
            (_b = (_a = host.getFormatOptions) === null || _a === void 0 ? void 0 : _a.call(host, document)) !== null && _b !== void 0 ? _b : {},
            (_d = (_c = host.getPreferences) === null || _c === void 0 ? void 0 : _c.call(host, document)) !== null && _d !== void 0 ? _d : {},
        ]);
        const fileName = shared.uriToFsPath(document.uri);
        const start = document.offsetAt(range.start);
        const end = document.offsetAt(range.end);
        let result = [];
        for (const error of context.diagnostics) {
            try {
                const codeFixes = languageService.getCodeFixesAtPosition(fileName, document.offsetAt(error.range.start), document.offsetAt(error.range.end), [Number(error.code)], formatOptions, preferences);
                for (const codeFix of codeFixes) {
                    result = result.concat(transformCodeFix(codeFix, [error]));
                }
            }
            catch { }
        }
        if (context.only) {
            for (const only of context.only) {
                if (only.split('.')[0] === vscode.CodeActionKind.Refactor) {
                    try {
                        const refactors = languageService.getApplicableRefactors(fileName, { pos: start, end: end }, preferences, undefined, only);
                        for (const refactor of refactors) {
                            result = result.concat(transformRefactor(refactor));
                        }
                    }
                    catch { }
                }
            }
        }
        else {
            try {
                const refactors = languageService.getApplicableRefactors(fileName, { pos: start, end: end }, preferences, undefined, undefined);
                for (const refactor of refactors) {
                    result = result.concat(transformRefactor(refactor));
                }
            }
            catch { }
        }
        if (matchOnlyKind(vscode.CodeActionKind.SourceOrganizeImports)) {
            const action = vscode.CodeAction.create('Organize Imports', vscode.CodeActionKind.SourceOrganizeImports);
            const data = {
                type: 'organizeImports',
                uri,
                fileName,
            };
            action.data = data;
            result.push(action);
        }
        if (matchOnlyKind(`${vscode.CodeActionKind.SourceFixAll}.ts`)) {
            const action = vscode.CodeAction.create('Fix All', vscode.CodeActionKind.SourceFixAll);
            const data = {
                uri,
                type: 'fixAll',
                fileName,
                fixIds: [
                    fixNames.classIncorrectlyImplementsInterface,
                    fixNames.awaitInSyncFunction,
                    fixNames.unreachableCode,
                ],
            };
            action.data = data;
            result.push(action);
        }
        if (matchOnlyKind(`${vscode.CodeActionKind.Source}.removeUnused.ts`)) {
            const action = vscode.CodeAction.create('Remove all unused code', vscode.CodeActionKind.SourceFixAll);
            const data = {
                uri,
                type: 'fixAll',
                fileName,
                fixIds: [
                    // not working and throw
                    fixNames.unusedIdentifier,
                    // TODO: remove patching
                    'unusedIdentifier_prefix',
                    'unusedIdentifier_deleteImports',
                    'unusedIdentifier_delete',
                    'unusedIdentifier_infer',
                ],
            };
            action.data = data;
            result.push(action);
        }
        if (matchOnlyKind(`${vscode.CodeActionKind.Source}.addMissingImports.ts`)) {
            const action = vscode.CodeAction.create('Add all missing imports', vscode.CodeActionKind.SourceFixAll);
            const data = {
                uri,
                type: 'fixAll',
                fileName,
                fixIds: [
                    // not working and throw
                    fixNames.fixImport,
                    // TODO: remove patching
                    'fixMissingImport',
                ],
            };
            action.data = data;
            result.push(action);
        }
        for (const codeAction of result) {
            if (codeAction.diagnostics === undefined) {
                codeAction.diagnostics = context.diagnostics;
            }
        }
        return result;
        function matchOnlyKind(kind) {
            if (context.only) {
                for (const only of context.only) {
                    const a = only.split('.');
                    const b = kind.split('.');
                    if (a.length <= b.length) {
                        let matchNums = 0;
                        for (let i = 0; i < a.length; i++) {
                            if (a[i] == b[i]) {
                                matchNums++;
                            }
                        }
                        if (matchNums === a.length)
                            return true;
                    }
                }
            }
        }
        function transformCodeFix(codeFix, diagnostics) {
            const edit = rename_1.fileTextChangesToWorkspaceEdit(codeFix.changes, getTextDocument);
            const codeActions = [];
            const fix = vscode.CodeAction.create(codeFix.description, edit, vscode.CodeActionKind.QuickFix);
            fix.diagnostics = diagnostics;
            codeActions.push(fix);
            if (codeFix.fixAllDescription && codeFix.fixId) {
                const fixAll = vscode.CodeAction.create(codeFix.fixAllDescription, vscode.CodeActionKind.QuickFix);
                const data = {
                    uri,
                    type: 'fixAll',
                    fileName,
                    fixIds: [codeFix.fixId],
                };
                fixAll.data = data;
                fixAll.diagnostics = diagnostics;
                codeActions.push(fixAll);
            }
            return codeActions;
        }
        function transformRefactor(refactor) {
            const codeActions = [];
            for (const action of refactor.actions) {
                const codeAction = vscode.CodeAction.create(action.description, action.kind);
                const data = {
                    uri,
                    type: 'refactor',
                    fileName,
                    range: { pos: start, end: end },
                    refactorName: refactor.name,
                    actionName: action.name,
                };
                codeAction.data = data;
                if (action.notApplicableReason) {
                    codeAction.disabled = { reason: action.notApplicableReason };
                }
                if (refactor.inlineable) {
                    codeAction.isPreferred = true;
                }
                codeActions.push(codeAction);
            }
            return codeActions;
        }
    };
}
exports.register = register;
