// vscode-vue-languageservice/out/services/callHierarchy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const shared = require("@volar/shared");
const upath = require("upath");
const dedupe = require("../utils/dedupe");
function register({ sourceFiles, getTsLs }) {
    function doPrepare(uri, position) {
        let vueItems = [];
        for (const tsLoc of sourceFiles.toTsLocations(uri, position)) {
            if (tsLoc.type === 'embedded-ts' && !tsLoc.range.data.capabilities.references)
                continue;
            const items = worker(tsLoc.lsType, tsLoc.uri, tsLoc.range.start);
            for (const item of items) {
                const data = {
                    lsType: tsLoc.lsType,
                    tsData: item.data,
                };
                item.data = data;
            }
            vueItems = vueItems.concat(items);
        }
        return dedupe.withLocations(vueItems);
    }
    function getIncomingCalls(item) {
        var _a;
        const data = item.data;
        const lsType = (_a = data === null || data === void 0 ? void 0 : data.lsType) !== null && _a !== void 0 ? _a : 'template';
        const tsLs = getTsLs(lsType);
        const tsItems = tsTsCallHierarchyItem(item);
        const tsIncomingItems = tsItems.map(tsLs.callHierarchy.getIncomingCalls).flat();
        const vueIncomingItems = [];
        for (const tsIncomingItem of tsIncomingItems) {
            const vueResult = toVueCallHierarchyItem(lsType, tsIncomingItem.from, tsIncomingItem.fromRanges);
            if (!vueResult)
                continue;
            const [vueItem, vueRanges] = vueResult;
            vueIncomingItems.push({
                from: vueItem,
                fromRanges: vueRanges,
            });
        }
        return dedupe.withCallHierarchyIncomingCalls(vueIncomingItems);
    }
    function getOutgoingCalls(item) {
        var _a;
        const data = item.data;
        const lsType = (_a = data === null || data === void 0 ? void 0 : data.lsType) !== null && _a !== void 0 ? _a : 'template';
        const tsLs = getTsLs(lsType);
        const tsItems = tsTsCallHierarchyItem(item);
        const tsIncomingItems = tsItems.map(tsLs.callHierarchy.getOutgoingCalls).flat();
        const vueIncomingItems = [];
        for (const tsIncomingItem of tsIncomingItems) {
            const vueResult = toVueCallHierarchyItem(lsType, tsIncomingItem.to, tsIncomingItem.fromRanges);
            if (!vueResult)
                continue;
            const [vueItem, vueRanges] = vueResult;
            vueIncomingItems.push({
                to: vueItem,
                fromRanges: vueRanges,
            });
        }
        return dedupe.withCallHierarchyOutgoingCalls(vueIncomingItems);
    }
    return {
        doPrepare,
        getIncomingCalls,
        getOutgoingCalls,
    };
    function worker(lsType, tsDocUri, tsPos) {
        const vueOrTsItems = [];
        const tsLs = getTsLs(lsType);
        const tsItems = tsLs.callHierarchy.doPrepare(tsDocUri, tsPos);
        for (const tsItem of tsItems) {
            const result = toVueCallHierarchyItem(lsType, tsItem, []);
            if (!result)
                continue;
            const [vueItem] = result;
            if (vueItem) {
                vueOrTsItems.push(vueItem);
            }
        }
        for (const item of vueOrTsItems) {
            if (!item.data)
                item.data = {};
            item.data.uri = tsDocUri;
        }
        return vueOrTsItems;
    }
    function toVueCallHierarchyItem(lsType, tsItem, tsRanges) {
        if (!sourceFiles.getTsDocuments(lsType).get(tsItem.uri)) {
            return [tsItem, tsRanges]; // not virtual file
        }
        const sourceMap = sourceFiles.getTsSourceMaps(lsType).get(tsItem.uri);
        if (!sourceMap)
            return;
        let vueRange = sourceMap.getSourceRange(tsItem.range.start, tsItem.range.end);
        if (!vueRange) {
            // TODO: <script> range
            vueRange = {
                start: sourceMap.sourceDocument.positionAt(0),
                end: sourceMap.sourceDocument.positionAt(sourceMap.sourceDocument.getText().length),
            };
        }
        const vueSelectionRange = sourceMap.getSourceRange(tsItem.selectionRange.start, tsItem.selectionRange.end);
        if (!vueSelectionRange)
            return;
        const vueRanges = tsRanges.map(tsRange => sourceMap.getSourceRange(tsRange.start, tsRange.end)).filter(shared.notEmpty);
        const vueItem = {
            ...tsItem,
            name: tsItem.name === upath.basename(shared.uriToFsPath(sourceMap.mappedDocument.uri)) ? upath.basename(shared.uriToFsPath(sourceMap.sourceDocument.uri)) : tsItem.name,
            uri: sourceMap.sourceDocument.uri,
            range: vueRange,
            selectionRange: vueSelectionRange,
        };
        return [vueItem, vueRanges];
    }
    function tsTsCallHierarchyItem(item) {
        if (upath.extname(item.uri) !== '.vue') {
            return [item];
        }
        const tsItems = [];
        for (const tsLoc of sourceFiles.toTsLocations(item.uri, item.range.start, item.range.end)) {
            if (tsLoc.type === 'embedded-ts' && !tsLoc.range.data.capabilities.references)
                continue;
            for (const tsSelectionLoc of sourceFiles.toTsLocations(item.uri, item.selectionRange.start, item.selectionRange.end)) {
                if (tsSelectionLoc.type === 'embedded-ts' && !tsSelectionLoc.range.data.capabilities.references)
                    continue;
                tsItems.push({
                    ...item,
                    uri: tsLoc.uri,
                    range: tsLoc.range,
                    selectionRange: tsSelectionLoc.range,
                });
            }
        }
        return tsItems;
    }
}
exports.register = register;
