// vscode-vue-languageservice/out/services/codeAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const transforms_1 = require("@volar/transforms");
const vscode = require("vscode-languageserver");
const dedupe = require("../utils/dedupe");
const rename_1 = require("./rename");
function register({ sourceFiles, getCssLs, getTsLs }) {
    return async (uri, range, context) => {
        const tsResult = await onTs(uri, range, context);
        const cssResult = onCss(uri, range, context);
        return dedupe.withCodeAction([
            ...tsResult,
            ...cssResult,
        ]);
    };
    async function onTs(uri, range, context) {
        let result = [];
        for (const tsLoc of sourceFiles.toTsLocations(uri, range.start, range.end)) {
            const tsLs = getTsLs(tsLoc.lsType);
            const tsContext = {
                diagnostics: transforms_1.transformLocations(context.diagnostics, vueRange => tsLoc.type === 'embedded-ts' ? tsLoc.sourceMap.getMappedRange(vueRange.start, vueRange.end) : vueRange),
                only: context.only,
            };
            if (tsLoc.type === 'embedded-ts' && !tsLoc.sourceMap.capabilities.codeActions)
                continue;
            let tsCodeActions = await tsLs.getCodeActions(tsLoc.uri, tsLoc.range, tsContext);
            if (!tsCodeActions)
                continue;
            for (const tsCodeAction of tsCodeActions) {
                if (tsCodeAction.title.indexOf('__VLS_') >= 0)
                    continue;
                const vueEdit = tsCodeAction.edit ? rename_1.tsEditToVueEdit(tsLoc.lsType, tsCodeAction.edit, sourceFiles, () => true) : undefined;
                if (tsCodeAction.edit && !vueEdit)
                    continue;
                const data = {
                    lsType: tsLoc.lsType,
                    tsData: tsCodeAction.data,
                };
                result.push({
                    ...tsCodeAction,
                    data,
                    edit: vueEdit,
                });
            }
        }
        return result;
    }
    function onCss(uri, range, context) {
        let result = [];
        const sourceFile = sourceFiles.get(uri);
        if (!sourceFile)
            return result;
        for (const sourceMap of sourceFile.getCssSourceMaps()) {
            if (!sourceMap.stylesheet)
                continue;
            const cssLs = getCssLs(sourceMap.mappedDocument.languageId);
            if (!cssLs)
                continue;
            for (const cssRange of sourceMap.getMappedRanges(range.start, range.end)) {
                const cssContext = {
                    diagnostics: transforms_1.transformLocations(context.diagnostics, vueRange => sourceMap.getMappedRange(vueRange.start, vueRange.end)),
                    only: context.only,
                };
                const cssCodeActions = cssLs.doCodeActions2(sourceMap.mappedDocument, cssRange, cssContext, sourceMap.stylesheet);
                for (const codeAction of cssCodeActions) {
                    // TODO
                    // cssCodeAction.edit?.changeAnnotations
                    // cssCodeAction.edit?.documentChanges...
                    if (codeAction.edit) {
                        const vueEdit = {};
                        for (const cssUri in codeAction.edit.changes) {
                            if (cssUri === sourceMap.mappedDocument.uri) {
                                if (!vueEdit.changes) {
                                    vueEdit.changes = {};
                                }
                                vueEdit.changes[uri] = transforms_1.transformLocations(vueEdit.changes[cssUri], cssRange_2 => sourceMap.getSourceRange(cssRange_2.start, cssRange_2.end));
                            }
                        }
                        if (codeAction.edit.documentChanges) {
                            for (const cssDocChange of codeAction.edit.documentChanges) {
                                if (!vueEdit.documentChanges) {
                                    vueEdit.documentChanges = [];
                                }
                                if (vscode.TextDocumentEdit.is(cssDocChange)) {
                                    cssDocChange.textDocument = {
                                        uri: uri,
                                        version: sourceMap.sourceDocument.version,
                                    };
                                    cssDocChange.edits = transforms_1.transformLocations(cssDocChange.edits, cssRange_2 => sourceMap.getSourceRange(cssRange_2.start, cssRange_2.end));
                                    vueEdit.documentChanges.push(cssDocChange);
                                }
                            }
                        }
                        codeAction.edit = vueEdit;
                    }
                    if (codeAction.diagnostics) {
                        codeAction.diagnostics = transforms_1.transformLocations(codeAction.diagnostics, cssRange_2 => sourceMap.getSourceRange(cssRange_2.start, cssRange_2.end));
                    }
                    result.push(codeAction);
                }
            }
        }
        return result;
    }
}
exports.register = register;
