// vscode-vue-languageservice/out/services/semanticTokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = exports.getSemanticTokenLegend = void 0;
const shared_1 = require("@vue/shared");
const vscode = require("vscode-languageserver");
const ts2 = require("vscode-typescript-languageservice"); // TODO: remove it
function getSemanticTokenLegend() {
    const tsLegend = ts2.getSemanticTokenLegend();
    const tokenTypesLegend = [
        ...tsLegend.types,
        'componentTag',
    ];
    const semanticTokenLegend = {
        tokenTypes: tokenTypesLegend,
        tokenModifiers: tsLegend.modifiers,
    };
    return semanticTokenLegend;
}
exports.getSemanticTokenLegend = getSemanticTokenLegend;
function register({ sourceFiles, getTsLs, htmlLs, pugLs, modules: { html } }, updateTemplateScripts) {
    const semanticTokensLegend = getSemanticTokenLegend();
    const tokenTypes = new Map(semanticTokensLegend.tokenTypes.map((t, i) => [t, i]));
    return (uri, range, cancle, resultProgress) => {
        const sourceFile = sourceFiles.get(uri);
        if (!sourceFile)
            return;
        const document = sourceFile.getTextDocument();
        const offsetRange = range ?
            {
                start: document.offsetAt(range.start),
                end: document.offsetAt(range.end),
            } : {
            start: 0,
            end: document.getText().length,
        };
        let tokens = [];
        if (cancle === null || cancle === void 0 ? void 0 : cancle.isCancellationRequested)
            return;
        const scriptResult = getTsResult(sourceFile, 'script');
        if (!(cancle === null || cancle === void 0 ? void 0 : cancle.isCancellationRequested) && scriptResult.length) {
            tokens = tokens.concat(scriptResult);
            resultProgress === null || resultProgress === void 0 ? void 0 : resultProgress.report(buildTokens(tokens));
        }
        if (sourceFile.getHtmlSourceMaps().length) {
            updateTemplateScripts();
        }
        if (cancle === null || cancle === void 0 ? void 0 : cancle.isCancellationRequested)
            return;
        const templateResult = getTsResult(sourceFile, 'template');
        if (!(cancle === null || cancle === void 0 ? void 0 : cancle.isCancellationRequested) && templateResult.length) {
            tokens = tokens.concat(templateResult);
            resultProgress === null || resultProgress === void 0 ? void 0 : resultProgress.report(buildTokens(tokens));
        }
        if (cancle === null || cancle === void 0 ? void 0 : cancle.isCancellationRequested)
            return;
        const htmlResult = getHtmlResult(sourceFile);
        if (!(cancle === null || cancle === void 0 ? void 0 : cancle.isCancellationRequested) && htmlResult.length) {
            tokens = tokens.concat(htmlResult);
            resultProgress === null || resultProgress === void 0 ? void 0 : resultProgress.report(buildTokens(tokens));
        }
        if (cancle === null || cancle === void 0 ? void 0 : cancle.isCancellationRequested)
            return;
        return buildTokens(tokens);
        function buildTokens(tokens) {
            var _a;
            const builder = new vscode.SemanticTokensBuilder();
            for (const token of tokens.sort((a, b) => a[0] - b[0] === 0 ? a[1] - b[1] : a[0] - b[0])) {
                builder.push(token[0], token[1], token[2], token[3], (_a = token[4]) !== null && _a !== void 0 ? _a : 0);
            }
            return builder.build();
        }
        function getTsResult(sourceFile, lsType) {
            const result = [];
            for (const sourceMap of sourceFile.getTsSourceMaps()) {
                if (sourceMap.lsType !== lsType)
                    continue;
                const tsLs = getTsLs(sourceMap.lsType);
                for (const maped of sourceMap) {
                    if (maped.data.capabilities.semanticTokens
                        && maped.sourceRange.end > offsetRange.start
                        && maped.sourceRange.start < offsetRange.end) {
                        if (cancle === null || cancle === void 0 ? void 0 : cancle.isCancellationRequested)
                            return result;
                        const tsRange = {
                            start: sourceMap.mappedDocument.positionAt(maped.mappedRange.start),
                            end: sourceMap.mappedDocument.positionAt(maped.mappedRange.end),
                        };
                        const tokens = tsLs.getDocumentSemanticTokens(sourceMap.mappedDocument.uri, tsRange, cancle);
                        if (!tokens)
                            continue;
                        for (const token of tokens) {
                            const tsStart = sourceMap.mappedDocument.offsetAt({ line: token[0], character: token[1] });
                            const tsEnd = sourceMap.mappedDocument.offsetAt({ line: token[0], character: token[1] + token[2] });
                            const vueRange = sourceMap.getSourceRange2(tsStart, tsEnd);
                            if (!vueRange || !vueRange.data.capabilities.semanticTokens)
                                continue;
                            const vuePos = document.positionAt(vueRange.start);
                            result.push([vuePos.line, vuePos.character, vueRange.end - vueRange.start, token[3], token[4]]);
                        }
                    }
                }
            }
            return result;
        }
        function getHtmlResult(sourceFile) {
            var _a;
            const result = [];
            const templateScriptData = sourceFile.getTemplateScriptData();
            const components = new Set([
                ...templateScriptData.components,
                ...templateScriptData.components.map(shared_1.hyphenate).filter(name => !shared_1.isHTMLTag(name)),
            ]);
            for (const sourceMap of [...sourceFile.getHtmlSourceMaps(), ...sourceFile.getPugSourceMaps()]) {
                const docText = sourceMap.mappedDocument.getText();
                const scanner = sourceMap.language === 'html'
                    ? htmlLs.createScanner(docText, offsetRange.start)
                    : pugLs.createScanner(sourceMap.pugDocument, offsetRange.start);
                if (!scanner)
                    continue;
                let token = scanner.scan();
                while (token !== html.TokenType.EOS) {
                    if (token === html.TokenType.StartTag || token === html.TokenType.EndTag) {
                        const tokenText = scanner.getTokenText();
                        if (components.has(tokenText)) {
                            const tokenOffset = scanner.getTokenOffset();
                            const tokenLength = scanner.getTokenLength();
                            const vueRange = sourceMap.getSourceRange2(tokenOffset);
                            if (vueRange) {
                                const vueOffset = vueRange.start;
                                if (vueOffset > offsetRange.end)
                                    break; // TODO: fix source map perf and break in while condition
                                const vuePos = sourceMap.sourceDocument.positionAt(vueOffset);
                                result.push([vuePos.line, vuePos.character, tokenLength, (_a = tokenTypes.get('componentTag')) !== null && _a !== void 0 ? _a : -1, undefined]);
                            }
                        }
                    }
                    token = scanner.scan();
                }
            }
            return result;
        }
    };
}
exports.register = register;
