--- postgresql-7.1beta4/src/pl/plperl/Makefile.PL.orig  Wed Jan 31 00:54:56 2001
+++ postgresql-7.1beta4/src/pl/plperl/Makefile.PL       Wed Jan 31 01:06:50 2001
@@ -10,5 +10,5 @@
 # seems pretty unlikely anyway.)
 #
-if ($Config{'useshrplib'} ne 'true') {
+if ($Config{'useshrplib'} ne 'true' && $ARGV[0] ne '--force') {
        open(OUT, ">Makefile") or die "Can't write Makefile: $!\n";
        print OUT <<'EndOfMakefile';

