// vscode-vue-languageservice/out/commands/unuseRefSugar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const shared = require("@volar/shared");
const vscode = require("vscode-languageserver");
const codeAction = require("../services/codeAction");
const codeActionResolve = require("../services/codeActionResolve");
const diagnostics = require("../services/diagnostics");
const rename = require("../services/rename");
function register(context) {
    const { sourceFiles } = context;
    const doRename = rename.register(context).doRename;
    const getDiagnostics = diagnostics.register(context, () => { });
    const getCodeActions = codeAction.register(context);
    const resolveCodeAction = codeActionResolve.register(context);
    return async (connection, uri) => {
        var _a;
        const sourceFile = sourceFiles.get(uri);
        if (!sourceFile)
            return;
        const descriptor = sourceFile.getDescriptor();
        if (!descriptor.scriptSetup)
            return;
        const scriptSetupAst = sourceFile.getScriptSetupAst();
        if (!scriptSetupAst)
            return;
        const progress = await connection.window.createWorkDoneProgress();
        progress.begin('Unuse Ref Sugar', 0, '', true);
        const edits = await getUnRefSugarEdits(sourceFile, descriptor.scriptSetup, scriptSetupAst);
        if (progress.token.isCancellationRequested)
            return;
        if (edits === null || edits === void 0 ? void 0 : edits.length) {
            await connection.workspace.applyEdit({ changes: { [uri]: edits } });
            await shared.sleep(0);
            const errors = (_a = await getDiagnostics(uri, () => { })) !== null && _a !== void 0 ? _a : [];
            await shared.sleep(0);
            const importEdits = await getAddMissingImportsEdits(sourceFile, descriptor.scriptSetup, errors);
            const removeInvalidValueEdits = getRemoveInvalidDotValueEdits(sourceFile, errors);
            if (importEdits && removeInvalidValueEdits) {
                rename.margeWorkspaceEdits(importEdits, removeInvalidValueEdits);
                await connection.workspace.applyEdit(importEdits);
            }
            else if (removeInvalidValueEdits) {
                await connection.workspace.applyEdit(removeInvalidValueEdits);
            }
        }
        progress.done();
        async function getAddMissingImportsEdits(_sourceFile, _scriptSetup, errors) {
            const document = _sourceFile.getTextDocument();
            const codeActions = await getCodeActions(uri, {
                start: document.positionAt(_scriptSetup.loc.start),
                end: document.positionAt(_scriptSetup.loc.start),
            }, {
                diagnostics: errors.filter(error => error.code === 2552),
                only: [`${vscode.CodeActionKind.Source}.addMissingImports.ts`],
            });
            for (const codeAction of codeActions) {
                const newCodeAction = await resolveCodeAction(codeAction);
                if (newCodeAction.edit) {
                    return newCodeAction.edit;
                }
            }
        }
        function getRemoveInvalidDotValueEdits(_sourceFile, errors) {
            const document = _sourceFile.getTextDocument();
            const edits = [];
            for (const error of errors) {
                const errorText = document.getText(error.range);
                if (error.code === 2339 && errorText === 'value') {
                    edits.push(vscode.TextEdit.del({
                        start: {
                            line: error.range.start.line,
                            character: error.range.start.character - 1,
                        },
                        end: error.range.end,
                    }));
                }
            }
            if (!edits.length)
                return;
            const result = { documentChanges: [vscode.TextDocumentEdit.create(vscode.OptionalVersionedTextDocumentIdentifier.create(uri, document.version), edits)] };
            return result;
        }
        async function getUnRefSugarEdits(_sourceFile, _scriptSetup, _scriptSetupAst) {
            const ranges = _sourceFile.getSfcRefSugarRanges();
            const document = _sourceFile.getTextDocument();
            const edits = [];
            if (!ranges)
                return;
            let varsNum = 0;
            let varsCur = 0;
            for (const callRange of ranges.refs) {
                varsNum += callRange.leftBindings.length;
            }
            for (const callRange of ranges.refs) {
                addReplace(callRange.flag.start, callRange.flag.end, 'const');
                const fnName = _scriptSetup.content.substring(callRange.rightFn.start, callRange.rightFn.end);
                if (fnName === '$fromRefs') {
                }
                else {
                    const newFnName = fnName.substring(1); // $ref -> ref
                    addReplace(callRange.rightFn.start, callRange.rightFn.end, newFnName);
                }
                for (const binding of callRange.leftBindings) {
                    if (progress === null || progress === void 0 ? void 0 : progress.token.isCancellationRequested)
                        return;
                    const varText = _scriptSetup.content.substring(binding.start, binding.end);
                    progress === null || progress === void 0 ? void 0 : progress.report(++varsCur / varsNum * 100, varText);
                    await shared.sleep(0);
                    const bindingName = _scriptSetup.content.substring(binding.start, binding.end);
                    const renames = await doRename(uri, document.positionAt(_scriptSetup.loc.start + binding.end), bindingName + '.value');
                    if (renames === null || renames === void 0 ? void 0 : renames.changes) {
                        const edits_2 = renames.changes[uri];
                        if (edits_2) {
                            for (const edit of edits_2) {
                                const editRange = {
                                    start: document.offsetAt(edit.range.start),
                                    end: document.offsetAt(edit.range.end),
                                };
                                if (editRange.start >= (_scriptSetup.loc.start + binding.start) && editRange.end <= (_scriptSetup.loc.start + binding.end))
                                    continue;
                                if (editRange.end < _scriptSetup.loc.start || editRange.start > _scriptSetup.loc.end)
                                    continue;
                                if (inRawCall(editRange.start, editRange.end))
                                    continue;
                                edits.push(edit);
                            }
                        }
                    }
                }
            }
            for (const rawCall of ranges.raws) {
                addReplace(rawCall.fullRange.start, rawCall.argsRange.start, '');
                addReplace(rawCall.argsRange.end, rawCall.fullRange.end, '');
            }
            return edits;
            function inRawCall(start, end) {
                if (ranges) {
                    for (const rawRange of ranges.raws) {
                        if (start >= (_scriptSetup.loc.start + rawRange.argsRange.start) && end <= (_scriptSetup.loc.start + rawRange.argsRange.end)) {
                            return true;
                        }
                    }
                }
                return false;
            }
            function addReplace(start, end, text) {
                if (_scriptSetup.content.substring(start, end) === text)
                    return;
                edits.push(vscode.TextEdit.replace({
                    start: document.positionAt(_scriptSetup.loc.start + start),
                    end: document.positionAt(_scriptSetup.loc.start + end),
                }, text));
            }
        }
    };
}
exports.register = register;
