// vscode-vue-languageservice/out/commands/useSetupSugar.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = void 0;
const shared = require("@volar/shared");
const vscode = require("vscode-languageserver");
const scriptSetupConvertRanges_1 = require("../parsers/scriptSetupConvertRanges");
const codeAction = require("../services/codeAction");
const codeActionResolve = require("../services/codeActionResolve");
const diagnostics = require("../services/diagnostics");
function register(context) {
    const { sourceFiles, modules: { typescript: ts } } = context;
    const getDiagnostics = diagnostics.register(context, () => { });
    const getCodeActions = codeAction.register(context);
    const resolveCodeAction = codeActionResolve.register(context);
    return async (connection, uri) => {
        var _a;
        const sourceFile = sourceFiles.get(uri);
        if (!sourceFile)
            return;
        const descriptor = sourceFile.getDescriptor();
        if (!descriptor.script)
            return;
        const scriptAst = sourceFile.getScriptAst();
        if (!scriptAst)
            return;
        const edits = await getEdits(sourceFile, descriptor.script, scriptAst);
        if (edits === null || edits === void 0 ? void 0 : edits.length) {
            await connection.workspace.applyEdit({ changes: { [uri]: edits } });
            await shared.sleep(0);
            const errors = (_a = await getDiagnostics(uri, () => { })) !== null && _a !== void 0 ? _a : [];
            await shared.sleep(0);
            const importEdits = await getAddMissingImportsEdits(sourceFile, descriptor.scriptSetup, errors);
            if (importEdits) {
                await connection.workspace.applyEdit(importEdits);
            }
        }
        async function getAddMissingImportsEdits(_sourceFile, _scriptSetup, errors) {
            const document = _sourceFile.getTextDocument();
            const codeActions = await getCodeActions(uri, {
                start: document.positionAt(_scriptSetup.loc.start),
                end: document.positionAt(_scriptSetup.loc.start),
            }, {
                diagnostics: errors.filter(error => error.code === 2552),
                only: [`${vscode.CodeActionKind.Source}.addMissingImports.ts`],
            });
            for (const codeAction of codeActions) {
                const newCodeAction = await resolveCodeAction(codeAction);
                if (newCodeAction.edit) {
                    return newCodeAction.edit;
                }
            }
        }
        async function getEdits(_sourceFile, _script, _scriptAst) {
            const ranges = scriptSetupConvertRanges_1.parseUseScriptSetupRanges(ts, _scriptAst);
            const document = _sourceFile.getTextDocument();
            const edits = [];
            const scriptStartPos = document.positionAt(_script.loc.start);
            const startTagText = document.getText({
                start: {
                    line: scriptStartPos.line,
                    character: 0,
                },
                end: scriptStartPos,
            });
            addReplace(-1, -1, ' setup');
            const newScriptSetupCode = getScriptSetupCode();
            const newScriptCode = getScriptCode();
            let newBlockCode = '\n' + newScriptSetupCode + '\n';
            if (newScriptCode !== '') {
                newBlockCode += '</script>\n';
                newBlockCode += '\n';
                newBlockCode += startTagText + '\n';
                newBlockCode += newScriptCode + '\n';
            }
            addReplace(0, _script.content.length, newBlockCode);
            return edits;
            function getScriptCode() {
                let scriptBodyCode = '';
                let scriptExportCode = '';
                for (const statement of ranges.otherScriptStatements) {
                    const statementRange = getStatementRange(statement);
                    scriptBodyCode += _script.content.substring(statementRange.start, statementRange.end) + '\n';
                }
                if (ranges.otherOptions.length) {
                    scriptExportCode += 'export default defineComponent({\n';
                    for (const otherOption of ranges.otherOptions) {
                        scriptExportCode += _script.content.substring(otherOption.start, otherOption.end) + ',\n';
                    }
                    scriptExportCode += '});\n';
                }
                return [scriptBodyCode, scriptExportCode]
                    .map(code => code.trim())
                    .filter(code => code !== '')
                    .join('\n\n');
            }
            function getScriptSetupCode() {
                var _a, _b, _c, _d, _e;
                let scriptSetupImportsCode = '';
                let scriptDefinesCode = '';
                let scriptSetupBodyCode = '';
                for (const importRange of ranges.imports) {
                    let importRange_2 = getStatementRange(importRange);
                    scriptSetupImportsCode += _script.content.substring(importRange_2.start, importRange_2.end) + '\n';
                }
                if (ranges.propsOption) {
                    if ((_a = ranges.setupOption) === null || _a === void 0 ? void 0 : _a.props) {
                        scriptDefinesCode += `const ${_script.content.substring(ranges.setupOption.props.start, ranges.setupOption.props.end)} = `;
                    }
                    scriptDefinesCode += `defineProps(${_script.content.substring(ranges.propsOption.start, ranges.propsOption.end)});\n`;
                }
                if (((_b = ranges.setupOption) === null || _b === void 0 ? void 0 : _b.context) && 'start' in ranges.setupOption.context) {
                    scriptDefinesCode += `const ${_script.content.substring(ranges.setupOption.context.start, ranges.setupOption.context.end)} = {\n`;
                    if (ranges.emitsOption) {
                        scriptDefinesCode += `emit: defineEmits(${_script.content.substring(ranges.emitsOption.start, ranges.emitsOption.end)}),\n`;
                    }
                    scriptDefinesCode += `slots: useSlots(),\n`;
                    scriptDefinesCode += `attrs: useAttrs(),\n`;
                    scriptDefinesCode += '};\n';
                }
                else {
                    if (ranges.emitsOption) {
                        if (((_c = ranges.setupOption) === null || _c === void 0 ? void 0 : _c.context) && 'emit' in ranges.setupOption.context && ranges.setupOption.context.emit) {
                            scriptDefinesCode += `const ${_script.content.substring(ranges.setupOption.context.emit.start, ranges.setupOption.context.emit.end)} = `;
                        }
                        scriptDefinesCode += `defineEmits(${_script.content.substring(ranges.emitsOption.start, ranges.emitsOption.end)});\n`;
                    }
                    if (((_d = ranges.setupOption) === null || _d === void 0 ? void 0 : _d.context) && 'slots' in ranges.setupOption.context && ranges.setupOption.context.slots) {
                        scriptDefinesCode += `const ${_script.content.substring(ranges.setupOption.context.slots.start, ranges.setupOption.context.slots.end)} = useSlots();\n`;
                    }
                    if (((_e = ranges.setupOption) === null || _e === void 0 ? void 0 : _e.context) && 'attrs' in ranges.setupOption.context && ranges.setupOption.context.attrs) {
                        scriptDefinesCode += `const ${_script.content.substring(ranges.setupOption.context.attrs.start, ranges.setupOption.context.attrs.end)} = useAttrs();\n`;
                    }
                }
                if (ranges.setupOption) {
                    const bodyRange = {
                        start: ranges.setupOption.body.start + 1,
                        end: ranges.setupOption.body.end - 1, // remove }
                    };
                    if (ranges.setupOption.bodyReturn) {
                        const reutrnRange = getStatementRange(ranges.setupOption.bodyReturn);
                        scriptSetupBodyCode = _script.content.substring(bodyRange.start, reutrnRange.start)
                            + _script.content.substring(reutrnRange.end, bodyRange.end);
                    }
                    else {
                        scriptSetupBodyCode = _script.content.substring(bodyRange.start, bodyRange.end);
                    }
                }
                return [scriptSetupImportsCode, scriptDefinesCode, scriptSetupBodyCode]
                    .map(code => code.trim())
                    .filter(code => code !== '')
                    .join('\n\n');
            }
            function getStatementRange(scriptTextRange) {
                let end = scriptTextRange.end;
                if (_script.content.substring(end, end + 1) === ';')
                    end++;
                return {
                    start: scriptTextRange.start,
                    end,
                };
            }
            function addReplace(start, end, text) {
                edits.push(vscode.TextEdit.replace({
                    start: document.positionAt(_script.loc.start + start),
                    end: document.positionAt(_script.loc.start + end),
                }, text));
            }
        }
    };
}
exports.register = register;
