// @volar/transforms/out/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformTextEdit = exports.transformSymbolInformations = exports.transformSymbolInformation = exports.transformSelectionRanges = exports.transformSelectionRange = exports.transformLocations = exports.transformLocation = exports.transformHover = exports.transformCompletionList = exports.transformCompletionItem = void 0;
var completionItem_1 = require("./completionItem");
Object.defineProperty(exports, "transformCompletionItem", { enumerable: true, get: function () { return completionItem_1.transform; } });
var completionList_1 = require("./completionList");
Object.defineProperty(exports, "transformCompletionList", { enumerable: true, get: function () { return completionList_1.transform; } });
var hover_1 = require("./hover");
Object.defineProperty(exports, "transformHover", { enumerable: true, get: function () { return hover_1.transform; } });
var locationLike_1 = require("./locationLike");
Object.defineProperty(exports, "transformLocation", { enumerable: true, get: function () { return locationLike_1.transform; } });
var locationsLike_1 = require("./locationsLike");
Object.defineProperty(exports, "transformLocations", { enumerable: true, get: function () { return locationsLike_1.transform; } });
var selectionRange_1 = require("./selectionRange");
Object.defineProperty(exports, "transformSelectionRange", { enumerable: true, get: function () { return selectionRange_1.transform; } });
var selectionRanges_1 = require("./selectionRanges");
Object.defineProperty(exports, "transformSelectionRanges", { enumerable: true, get: function () { return selectionRanges_1.transform; } });
var symbolInformation_1 = require("./symbolInformation");
Object.defineProperty(exports, "transformSymbolInformation", { enumerable: true, get: function () { return symbolInformation_1.transform; } });
var symbolInformations_1 = require("./symbolInformations");
Object.defineProperty(exports, "transformSymbolInformations", { enumerable: true, get: function () { return symbolInformations_1.transform; } });
var textEdit_1 = require("./textEdit");
Object.defineProperty(exports, "transformTextEdit", { enumerable: true, get: function () { return textEdit_1.transform; } });
