// vscode-pug-languageservice/out/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLanguageService = void 0;
const pugDocument_1 = require("./pugDocument");
const completion_1 = require("./services/completion");
const documentHighlight_1 = require("./services/documentHighlight");
const documentLinks_1 = require("./services/documentLinks");
const documentSymbol_1 = require("./services/documentSymbol");
const hover_1 = require("./services/hover");
const scanner_1 = require("./services/scanner");
const selectionRanges_1 = require("./services/selectionRanges");
function getLanguageService(htmlLs) {
    return {
        parsePugDocument: (doc) => pugDocument_1.parsePugDocument(doc, htmlLs),
        doComplete: completion_1.register(htmlLs),
        findDocumentHighlights: documentHighlight_1.register(htmlLs),
        findDocumentLinks: documentLinks_1.register(htmlLs),
        findDocumentSymbols: documentSymbol_1.register(htmlLs),
        doHover: hover_1.register(htmlLs),
        createScanner: scanner_1.register(htmlLs),
        getSelectionRanges: selectionRanges_1.register(htmlLs),
    };
}
exports.getLanguageService = getLanguageService;
