// vscode-vue-languageservice/out/use/useSfcJsons.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSfcJsons = void 0;
const vscode_languageserver_textdocument_1 = require("vscode-languageserver-textdocument");
const reactivity_1 = require("@vue/reactivity");
const SourceMaps = require("../utils/sourceMaps");
function useSfcJsons(getUnreactiveDoc, customBlocks, context) {
    let version = 0;
    const textDocuments = reactivity_1.computed(() => {
        const vueDoc = getUnreactiveDoc();
        const documents = [];
        for (let i = 0; i < customBlocks.value.length; i++) {
            const customBlock = customBlocks.value[i];
            const lang = customBlock.lang;
            const content = customBlock.content;
            const uri = vueDoc.uri + '.' + i + '.' + lang;
            const document = vscode_languageserver_textdocument_1.TextDocument.create(uri, lang, version++, content);
            if (lang === 'json' || lang === 'jsonc') {
                documents.push({
                    index: i,
                    textDocument: document,
                    jsonDocument: context.jsonLs.parseJSONDocument(document),
                });
            }
        }
        return documents;
    });
    const sourceMaps = reactivity_1.computed(() => {
        const vueDoc = getUnreactiveDoc();
        const sourceMaps = [];
        for (const doc of textDocuments.value) {
            const customBlock = customBlocks.value[doc.index];
            const sourceMap = new SourceMaps.JsonSourceMap(vueDoc, doc.textDocument, doc.jsonDocument);
            sourceMap.add({
                data: undefined,
                mode: SourceMaps.Mode.Offset,
                sourceRange: {
                    start: customBlock.loc.start,
                    end: customBlock.loc.end,
                },
                mappedRange: {
                    start: 0,
                    end: customBlock.loc.end - customBlock.loc.start,
                },
            });
            sourceMaps.push(sourceMap);
        }
        return sourceMaps;
    });
    return {
        textDocuments,
        sourceMaps,
    };
}
exports.useSfcJsons = useSfcJsons;
