// vscode-vue-languageservice/out/use/useSfcTemplateCompileResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSfcTemplateCompileResult = void 0;
const vscode = require("vscode-languageserver");
const reactivity_1 = require("@vue/reactivity");
const CompilerDOM = require("@vue/compiler-dom");
const CompilerVue2 = require("../utils/vue2TemplateCompiler");
function useSfcTemplateCompileResult(htmlDocument, isVue2Mode) {
    return reactivity_1.computed(() => {
        if (!htmlDocument.value)
            return;
        const errors = [];
        let ast;
        try {
            ast = (isVue2Mode ? CompilerVue2 : CompilerDOM).compile(htmlDocument.value.getText(), {
                onError: err => onCompilerError(err, vscode.DiagnosticSeverity.Error),
                onWarn: err => onCompilerError(err, vscode.DiagnosticSeverity.Warning),
            }).ast;
            function onCompilerError(err, severity) {
                if (!err.loc)
                    return;
                const diagnostic = {
                    range: {
                        start: htmlDocument.value.positionAt(err.loc.start.offset),
                        end: htmlDocument.value.positionAt(err.loc.end.offset),
                    },
                    severity,
                    code: err.code,
                    source: 'vue',
                    message: err.message,
                };
                errors.push(diagnostic);
            }
        }
        catch (err) {
            const diagnostic = {
                range: {
                    start: htmlDocument.value.positionAt(0),
                    end: htmlDocument.value.positionAt(htmlDocument.value.getText().length),
                },
                severity: vscode.DiagnosticSeverity.Error,
                code: err.code,
                source: 'vue',
                message: err.message,
            };
            errors.push(diagnostic);
        }
        return {
            errors,
            ast,
        };
    });
}
exports.useSfcTemplateCompileResult = useSfcTemplateCompileResult;
