diff -uNbr jakarta-tomcat-4.0.3-src/catalina/src/share/org/apache/catalina/connector/http/HttpResponseStream.java jakarta-tomcat-4.0.3-new/catalina/src/share/org/apache/catalina/connector/http/HttpResponseStream.java
--- jakarta-tomcat-4.0.3-src/catalina/src/share/org/apache/catalina/connector/http/HttpResponseStream.java	Fri Mar  1 17:49:54 2002
+++ jakarta-tomcat-4.0.3-new/catalina/src/share/org/apache/catalina/connector/http/HttpResponseStream.java	Thu Apr  4 10:42:09 2002
@@ -250,7 +250,7 @@
     protected void checkHead(HttpResponseImpl response) {
         HttpServletRequest servletRequest = 
             (HttpServletRequest) response.getRequest();
-        if (servletRequest.getMethod().equals("HEAD"))
+        if ("HEAD".equals(servletRequest.getMethod()))
             writeContent = false;
         else
             writeContent = true;
