diff --git a/features/alarms/history/rest/api/src/main/java/org/opennms/features/alarms/history/rest/api/AlarmHistoryRestService.java b/features/alarms/history/rest/api/src/main/java/org/opennms/features/alarms/history/rest/api/AlarmHistoryRestService.java
index 80ae3ffbb69..61a5bde349b 100644
--- a/features/alarms/history/rest/api/src/main/java/org/opennms/features/alarms/history/rest/api/AlarmHistoryRestService.java
+++ b/features/alarms/history/rest/api/src/main/java/org/opennms/features/alarms/history/rest/api/AlarmHistoryRestService.java
@@ -51,22 +51,24 @@ import org.opennms.features.alarms.history.api.AlarmState;
  * ?match-type=alarm-id   (default to =reduction-key)
  * ?at=3423842384834      (default to =$now)
  */
-@Path("alarm/")
+@Path("alarm/history")
 public interface AlarmHistoryRestService {
 
     @GET
-    @Path("history/{alarmId}/journal")
+    @Path("{alarmId}/journal")
     @Produces(MediaType.APPLICATION_JSON)
-    Collection<AlarmState> getStatesForAlarm(@PathParam("alarmId") String alarmId, @QueryParam("match-type") String matchType, @QueryParam("at") long time);
+    Collection<AlarmState> getStatesForAlarm(@PathParam("alarmId") String alarmId,
+            @QueryParam("match-type") String matchType);
 
     @GET
-    @Path("history/{alarmId}")
+    @Path("{alarmId}")
     @Produces(MediaType.APPLICATION_JSON)
-    Collection<AlarmState> getStatesForAlarmWithReductionKey(@PathParam("reductionKey") String reductionKey);
+    AlarmState getAlarm(@PathParam("alarmId") String alarmId,
+            @QueryParam("match-type") String matchType,
+            @QueryParam("at") Long time);
 
     @GET
-    @Path("history")
     @Produces(MediaType.APPLICATION_JSON)
-    Collection<AlarmState> getActiveAlarmsAt(@PathParam("timestampInMillis") long timestampInMillis);
+    Collection<AlarmState> getActiveAlarmsAt(@QueryParam("at") Long timestampInMillis);
 
 }
diff --git a/features/alarms/history/rest/impl/src/main/java/org/opennms/features/alarms/history/rest/impl/AlarmHistoryRestServiceImpl.java b/features/alarms/history/rest/impl/src/main/java/org/opennms/features/alarms/history/rest/impl/AlarmHistoryRestServiceImpl.java
index f5569cbcd4e..0a6a0b4de86 100644
--- a/features/alarms/history/rest/impl/src/main/java/org/opennms/features/alarms/history/rest/impl/AlarmHistoryRestServiceImpl.java
+++ b/features/alarms/history/rest/impl/src/main/java/org/opennms/features/alarms/history/rest/impl/AlarmHistoryRestServiceImpl.java
@@ -44,22 +44,26 @@ public class AlarmHistoryRestServiceImpl implements AlarmHistoryRestService {
     }
 
     @Override
-    public Collection<AlarmState> getStatesForAlarmWithDbId(int alarmId) {
-        return alarmHistoryRepository.getStatesForAlarmWithDbId(alarmId);
+    public Collection<AlarmState> getStatesForAlarm(String alarmId, String matchType) {
+        if ("alarm-id".equals(matchType)) {
+            return alarmHistoryRepository.getStatesForAlarmWithDbId(Integer.valueOf(alarmId));
+        }
+        return alarmHistoryRepository.getStatesForAlarmWithReductionKey(alarmId);
     }
 
     @Override
-    public Collection<AlarmState> getStatesForAlarmWithReductionKey(String reductionKey) {
-        return alarmHistoryRepository.getStatesForAlarmWithReductionKey(reductionKey);
+    public AlarmState getAlarm(String alarmId, String matchType, Long time) {
+        long timestampInMillis = time == null ? System.currentTimeMillis() : time;
+        if ("alarm-id".equals(matchType)) {
+            return alarmHistoryRepository.getAlarmWithDbIdAt(Integer.valueOf(alarmId), timestampInMillis);
+        }
+        return alarmHistoryRepository.getAlarmWithReductionKeyIdAt(alarmId, timestampInMillis);
     }
 
     @Override
-    public Collection<AlarmState> getActiveAlarmsAt(long timestampInMillis) {
+    public Collection<AlarmState> getActiveAlarmsAt(Long time) {
+        long timestampInMillis = time == null ? System.currentTimeMillis() : time;
         return alarmHistoryRepository.getActiveAlarmsAt(timestampInMillis);
     }
 
-    @Override
-    public Collection<AlarmState> getActiveAlarmsNow() {
-        return alarmHistoryRepository.getActiveAlarmsNow();
-    }
 }
